### PR TITLE
Added file parser and accompanied tests

### PR DESCRIPTION
**Pull request by gosreya**

### CODE REVIEW

File: Code/GraphMol/FileParsers/XYZFileParser.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    
+    return atom;
+}
+
+
+RWMol *XYZDataStreamToMol(std::istream &inStream) {
+    PRECONDITION(inStream, "no stream");
+    unsigned int numAtoms = 0;
+    
+    std::string_view num{getLine(inStream)};
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/XYZFileParser.cpp]: This pattern isn't safe: since the string_view doesn't own the memory it uses, you need to keep the lifetime of that memory in mind. In this case the temporary string returned by `getLine()` will be destroyed after executing this line, which will leave `num` pointing to deallocated memory. 

It's unfortunate that g++ doesn't warn about this (clang++ actually does).

There are two straightforward alternatives:
```
std::string line{getLine(inStream)};
std::string_view num{line};
```
or just use an std::string here, which is what I would recommend in this case:
```suggestion
    std::string num{getLine(inStream)};
```

File: Code/GraphMol/FileParsers/XYZFileParser.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        std::ostringstream errout;
+        errout << "Unable to recognize the number of atoms: cannot convert '" << num << "' to unsigned int on line 0" << std::endl;
+        throw FileParseException(errout.str());
+    }
+    
+    std::string comment{getLine(inStream)};
+    
+    RWMol *mol = nullptr;
+    if (numAtoms != 0) {
+        mol = new RWMol();
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/XYZFileParser.cpp]: As written, this will return a nullptr if there are zero atoms in the input. 

File formats like Mol/SDF allow zero-atom molecules and the RDKit supports that, so you might think about doing something like:
```suggestion
    RWMol *mol = new RWMol();
    if  (numAtoms) {
```


BTW: adding a test for this edge case would be a good idea.

File: Code/GraphMol/FileParsers/FileParsers.h [hunk 3]
```diff
@@ -286,6 +286,13 @@ RDKIT_FILEPARSERS_EXPORT RWMol *Mol2BlockToMol(
     const std::string &molBlock, bool sanitize = true, bool removeHs = true,
     Mol2Type variant = CORINA, bool cleanupSubstructures = true);
 
+// \brief construct a molecule from an xyz file
+/*!
+ *   \param fName    - string containing the file name
+ *   \param charge - the integer charge of the molecule
+ */
+RDKIT_FILEPARSERS_EXPORT RWMol *XYZFileToMol(const std::string &fName, int charge);
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/FileParsers.h]: You need to add declarations for the other two XYZ to mol functions

File: Code/GraphMol/FileParsers/XYZFileParser.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
+#include <boost/lexical_cast.hpp>
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/XYZFileParser.cpp:11]: Please add a copyright statement to the top of this and other new source files. I would suggest something like this:
```
//
//  Copyright (C) 2022 Sreya Gogineni and other RDKit contributors
//
//   @@ All Rights Reserved @@
//  This file is part of the RDKit.
//  The contents are covered by the terms of the BSD license
//  which is included in the file license.txt, found at the root
//  of the RDKit source tree.
//
```

File: Code/GraphMol/FileParsers/XYZFileParser.cpp [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <cstdlib>
+#include <cstdio>
+
+namespace RDKit {
+
+void ParseExtraLine(std::string_view extraLine) {
+    std::string_view whitespace{" \v\t"};
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/XYZFileParser.cpp]: huh, I've never seen `\v` before.

File: Code/GraphMol/FileParsers/XYZFileParser.cpp [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    if (symb.size() == 2 && symb[1] >= 'A' && symb[1] <= 'Z') {
+        symb[1] = static_cast<char>(tolower(symb[1]));
+    }
+    
+    try {
+      atom->setAtomicNum(PeriodicTable::getTable()->getAtomicNumber(symb));
+    } catch (const Invar::Invariant &e) {
+      delete atom;
+      throw FileParseException(e.what());
+    }
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/XYZFileParser.cpp]: This could be a bit simpler:
```suggestion
    std::string symb{atomLine.substr(delim0, delim1 - delim0)};
    if (symb.size() == 2 && symb[1] >= 'A' && symb[1] <= 'Z') {
        symb[1] = static_cast<char>(tolower(symb[1]));
    }
    
    Atom *atom;
    try {
      atom = new Atom(PeriodicTable::getTable()->getAtomicNumber(symb));
    } catch (const Invar::Invariant &e) {
      throw FileParseException(e.what());
    }
```

File: Code/GraphMol/FileParsers/XYZFileParser.cpp [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    
+    return atom;
+}
+
+
+RWMol *XYZDataStreamToMol(std::istream &inStream) {
+    PRECONDITION(inStream, "no stream");
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/XYZFileParser.cpp]: We normally use tests like this when passing in pointers. Since you are taking a reference, you don't need the `PRECONDITION`

File: Code/GraphMol/FileParsers/XYZFileParser.cpp [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        throw FileParseException(errout.str());
+    }
+    
+    std::string comment{getLine(inStream)};
+    
+    RWMol *mol = nullptr;
+    if (numAtoms != 0) {
+        mol = new RWMol();
+        Conformer *conf = new Conformer(numAtoms);
+        if (comment != "") {
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/XYZFileParser.cpp]:
```suggestion
        if (!comment.empty()) {
```

File: Code/GraphMol/FileParsers/XYZFileParser.cpp [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        for (unsigned int i = 0; i < numAtoms; i++) {
+            if (inStream.eof()) {
+                throw FileParseException("EOF hit while reading atoms");
+            }
+            RDGeom::Point3D pos;
+            std::string_view atomLine{getLine(inStream)};
+            Atom *atom = ParseXYZFileAtomLine(atomLine, pos, i + 2);
+            unsigned int idx = mol->addAtom(atom, false, true);
+            conf->setAtomPos(idx, pos);
+            mol->setAtomBookmark(atom, idx);
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/XYZFileParser.cpp]: Since you aren't using the bookmarks anywhere, you might as well avoid setting them.

File: Code/GraphMol/FileParsers/XYZFileParser.cpp [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+        }
+        mol->addConformer(conf);
+    }
+    
+    while (!inStream.eof()) {
+        std::string_view extraLine{getLine(inStream)};
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/XYZFileParser.cpp]: for the same reason discussed above:
```suggestion
        std::string extraLine{getLine(inStream)};
```

**Review by greglandrum — COMMENTED** [review]
Some comments and suggestions.

File: Code/GraphMol/FileParsers/XYZFileParser.cpp (hunk 1 above)
Comment (gosreya) [reply Code/GraphMol/FileParsers/XYZFileParser.cpp]: Ahh I see, thanks for the catch!

File: Code/GraphMol/FileParsers/XYZFileParser.cpp (hunk 5 above)
Comment (gosreya) [reply Code/GraphMol/FileParsers/XYZFileParser.cpp]: Yeah I found '\v' (vertical tab) in a list of whitespace characters and thought to include it since it wasn't a newline character --> I searched it up though and it appears to be kind of an outdated character that wouldn't make for an accurate xyz file line. Example I found:

The line 'print("hello\vworld");' would cause the output:
hello
        world

File: Code/GraphMol/FileParsers/FileParsers.h [hunk 11]
```diff
@@ -286,6 +286,17 @@ RDKIT_FILEPARSERS_EXPORT RWMol *Mol2BlockToMol(
     const std::string &molBlock, bool sanitize = true, bool removeHs = true,
     Mol2Type variant = CORINA, bool cleanupSubstructures = true);
 
+
+RDKIT_FILEPARSERS_EXPORT void ParseExtraLine(std::string extraLine);
+RDKIT_FILEPARSERS_EXPORT Atom *ParseXYZFileAtomLine(std::string atomLine, RDGeom::Point3D &pos, unsigned int line);
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/FileParsers.h]: Strings should generally be passed by const reference in order to avoid copying:
```suggestion
RDKIT_FILEPARSERS_EXPORT void ParseExtraLine(const std::string &extraLine);
RDKIT_FILEPARSERS_EXPORT Atom *ParseXYZFileAtomLine(const std::string &atomLine, RDGeom::Point3D &pos, unsigned int line);
```
you will need to update the definitions too.

Comment (greglandrum) [reply Code/GraphMol/FileParsers/FileParsers.h]: Also: unless the `ParseExtraLine()` function is being used in code outside of XYZParser.cpp, which doesn't seem to be the case, the function should not be in the public API. So you should remove that declaration from this file.

File: Code/GraphMol/FileParsers/FileParsers.h [hunk 12]
```diff
@@ -286,6 +286,17 @@ RDKIT_FILEPARSERS_EXPORT RWMol *Mol2BlockToMol(
...
+RDKIT_FILEPARSERS_EXPORT RWMol *XYZDataStreamToMol(std::istream &inStream);
+// \brief construct a molecule from an xyz file
+/*!
+ *   \param fName    - string containing the file name
+ *   \param charge - the integer charge of the molecule
+ */
+RDKIT_FILEPARSERS_EXPORT RWMol *XYZFileToMol(const std::string &fName, int charge=0);
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/FileParsers.h:302]: Still also need a declaration of `XYZBlockToMol` in order to be consistent with the other file parsers.

**Review by greglandrum — CHANGES_REQUESTED** [review]
a few more changes requested

**Review by greglandrum — APPROVED** [review]
LGTM!
